embassy-lora/src/lib.rs

```diff
@@ -7,6 +7,8 @@ pub(crate) mod fmt;
 
 #[cfg(feature = "stm32wl")]
 pub mod stm32wl;
+#[cfg(feature = "rak4631")]
```

Comment:
Feature should be named sx126x instead I think.

Reply:
- Creating sx12xx : 1) Semtech does not appear to have attempted consolidating the drivers for sx126x and sx127x; 2) Is it worth the migration pain for existing users of the sx127x driver?  I imagine developers will stick with the sx127x for existing products, but going forward will choose sx126x for improved battery performance, if given the option.  That said, https://github.com/Lora-net/LoRaMac-node/tree/master/src/radio would be a good place to evaluate the differences to be overcome in an sx12xx implementation.  I'll take a look.
- RadioSwitch:  Will look into it.  Thanks!
- The rak4631 feature was intended to indicate the flavor of sx126x package so that the correct implementation is used in board_specific.rs.  I will evaluate whether a new feature is needed to identify sx126x separately.

One thing I failed to mention in the initial post is that the GFSK aspect of the sx126x chip is not implemented in the current driver.  If a GFSK driver is needed, it seems like it should be implemented in a GFSK-specific portion of the Embassy code-base as a separate concern.

I greatly appreciate having other sets of eyes look at this.

Reply:
Thanks for looking into this!

1) The reason for my question was partially because I've used https://github.com/lthiery/rust-sx12xx/ wrapper before, and because it seems the chips have lot of registers in common, and lots of the settings are similar. But keeping them separate is also fine if you think that's better overall.
2) Since this is early stage and not even released on crates.io, we don't give any API guarantees just yet. Given the main purpose of embassy-lora to integrate with the lorawan-device crate, the API surface used in practice by users are ::new() to create the driver before wrapping it in lorawan-device. 

I've ignored GFSK in the sx127x driver, but I think there are some remains from the original driver it was based off. I think we can ignore/remove the GFSK part, as it is probably better served by the original sx127x driver anyway, but if you want to keep it that's fine too.

Reply:
As shown in Semtech's LORA repository (https://github.com/Lora-net/LoRaMac-node/tree/master/src/radio), the four different radio implementations (lr1110, sx126x, sx1272, and sx1276) mostly support the same LORA radio API.  It gets a little muddied in the sx1272 and sx1276 implementations, but the API is clear in the lr1110 and sx126x implementations:

Init
GetStatus
SetModem
SetChannel
IsChannelFree
Random
SetRxConfig
SetTxConfig
CheckRfFrequency
TimeOnAir
Send
Sleep
Standby
Rx
StartCad
SetTxContinuousWave
Rssi
Write
Read
WriteBuffer
ReadBuffer
SetMaxPayloadLength
SetPublicNetwork
GetWakeupTime
IrqProcess

Optional depending on radio:
RxBoosted,
SetRxDutyCycle

- In the implementation being evaluated here, embassy-lora/src/sx126x/sx126x_lora/mod.rs supports this interface, with a couple of exceptions related to GFSK (CheckRfFrequency and SetPublicNetwork).
- The file embassy-lora/src/sx126x/mod.rs provides the PhyRxTx trait and other plumbing to support existing Embassy LORAWAN functionality.

If we are to move forward with an sx12xx implementation and believe (as Semtech apparently does) that the LORA API above identifies a solid interface to LORA PHY functionality, the current Embassy sx127x implementation would need significant refactoring.  It is certainly feasible, but I am not sure if the implementation effort is warranted since all current radios support the PhyRxTx interface.

Reply:
I am unclear on the intended use of the RadioSwitch trait.  It would be great if some examples of potential use can be provided, so that I can determine how the current antenna handling can be improved.

On a related note, I just realized that output pins can also be degraded, as I did for input pins.  That might allow me to significantly lessen the generic jumbles like:

impl<SPI, CS, RESET, ANTRX, ANTTX, WAIT, BUS> Sx126xRadio<SPI, CS, RESET, ANTRX, ANTTX, WAIT, BUS>

CS, RESET, ANTRX, and WAIT can be condensed to one generic.  Let me know if I am generically jumbled myself.

Reply:
Agreed on letting the sx127x implementation stay as is for now maybe, but maybe someone who's willing can merge them in the future. Speaking of that, the existing STM32WL SubGhz radio, is actually a Sx126x as well (https://github.com/embassy-rs/embassy/blob/master/embassy-stm32/src/subghz/mod.rs), but with a 'special' SPI bus. Probably not worth merging that either, but it might serve as an implementation guide at least :)

On radio switch: By the looks of it, the ANTRX and ANTTX would be part of the RadioSwitch implementation, like this: https://github.com/embassy-rs/embassy/blob/master/examples/stm32wl/src/bin/lorawan.rs#L19

In the case of the sx127x, a board based on RAK811 needs a RadioSwitch implementation different to a board like the STM32 L0 Lora Discovery. But the driver stays the same, and does not need to provide feature flags for different boards.

The RadioSwitch is used in the embassy-lora to configure the antenna pins for TX or RX or sleep, like this https://github.com/embassy-rs/embassy/blob/master/embassy-lora/src/stm32wl/mod.rs#L134

Reply:
I agree that it would be beneficial to put all LORA radio physical layer processing under one architectural structure, regardless whether the driver for the radio is stm or nrf.  Furthermore, it would be good to provide a configuration mechanism - maybe features, maybe not - to associate generic LORA physical layer processing with specific processing for each type of Semtech chip.

The sx126x implementation provides a generic API, a set of generic subroutines, and a file to hold board-specific functionality chosen by feature.  I believe the generic API (listed in a previous comment) is a decent starting point, but there are multiple ways to bind specific functionality to this generic API, based on stm/nrf chip and Semtech chip.

I will be unable to do much additional LORA implementation until December; however, I would be interested in implementing a more cohesive LORA physical layer across chips.  If you would like to lay out an overall architecture in the interim, I would be willing to implement it.  My biggest deficiency is that I don't have the equipment to test chip combinations.  But I believe that with your guidance I can get the implementation close enough that we can bring it to fruition.

In the meantime, I will incorporate RadioSwitch processing for antennas and look at simplifying my use of generics.  I assume it is still reasonable to get the current sx126x implementation into the pot so folks can start turning up the heat on it.

Reply:
I would prefer if there was nothing board-specific, but if you're committed to improve that, I think we can rename the feature sx126x and fix the implementation to not be board-specific. 

I don't want board-specific feature flags to creep into embassy-lora unless it can be avoided, and if we add those flags it can start off in the wrong direction if people start adding more feature flags for different boards.